Increase the min supported version of the Zip spec

ianhays opened the pull request:
Zips created by ZipArchive right now say that they will support any version of the ZIP format. However, they will actually only support versions 2.0 and up. This would mean that zips created by the core version of ziparchive will be improperly unzipped by unzippers from versions 0.0 to 2.0 (1989 to 1993).

This commit fixes a regression from the full framework that sets the min to 2.0. The windows zip utility also sets the version to 2.0.

@ericstj

ianhays commented:
regression introduced in https://github.com/dotnet/corefx/commit/44a11c21ebb6c059995b64c8f38d3d14aef154d2 via https://github.com/dotnet/corefx/pull/4999

stephentoub commented:
LGTM

ericstj commented:
LGTM.

FWIW I also noticed another diff in that all entries had UnicodeFileName set, whereas the same ZIP produced by desktop would not.  I was able to workaround it by specifying ASCII encoding when creating the archive but that's something else to look into.

ianhays commented:
>FWIW I also noticed another diff in that all entries had UnicodeFileName set, whereas the same ZIP produced by desktop would not. I was able to workaround it by specifying ASCII encoding when creating the archive but that's something else to look into.

This is (for better or worse) working as intended:
```
        private Encoding DefaultSystemEncoding
        {
            // On the desktop, this was Encoding.GetEncoding(0), which gives you the encoding object
            // that corresponds too the default system codepage.
            // However, in ProjectN, not only Encoding.GetEncoding(Int32) is not exposed, but there is also
            // no guarantee that a notion of a default system code page exists on the OS.
            // In fact, we can really only rely on UTF8 and UTF16 being present on all platforms.
            // We fall back to UTF8 as this is what is used by ZIP when as the "unicode encoding".
            get
            {
                return Encoding.UTF8;
                // return Encoding.GetEncoding(0);
            }
        }
```

ericstj commented:
I can't imagine that ZIP wants things encoded in a portable file as `default system codepage`.  That comment is made in isolation of the actual code around it and its use.  If you check the ZIP spec on this the actual behavior is to treat these as codepage 437 (see Appendix D of https://pkware.cachefly.net/webdocs/casestudies/APPNOTE.TXT) .  That codepage includes ASCII 0-126 so our [check for Unicode](https://github.com/dotnet/corefx/blob/d0dc5fc099946adc1035b34a8b1f6042eddb0c75/src/System.IO.Compression/src/System/IO/Compression/ZipHelper.cs#L22) would be sufficient to switch between ASCII and UTF8.

ianhays commented:
>I can't imagine that ZIP wants things encoded in a portable file as default system codepage.

This is fine as long as the default system codepage is either UTF8 or a non-UTF8 encoding that conforms to the old spec and as long as we set the Unicode bit accordingly. I think that's what the comment is trying to say: it sets the encoding to UTF8 to avoid any potential encoding issues.

>That codepage includes ASCII 0-126 so our check for Unicode would be sufficient to switch between ASCII and UTF8.

Yes, we could do that. Are you suggesting we only set to ASCII when that is the current encoding, or are you suggesting we set to ASCII for all cases where the string doesn't require Unicode?

ericstj commented:
Actually, it looks like this is only used in the decode phase.  I'm interested in the encode side which seems to be broken.  We can take that discussion to a separate issue.

ianhays commented:
>Actually, it looks like this is only used in the decode phase. I'm interested in the encode side which seems to be broken. 

[It's the Encoding that uses the RequriesUnicode method](https://github.com/dotnet/corefx/blob/d0dc5fc099946adc1035b34a8b1f6042eddb0c75/src/System.IO.Compression/src/System/IO/Compression/ZipArchiveEntry.cs#L474) and both encoding/decoding call the DefaultSystemEncoding. The problem is that because DefaultSystemEncoding is hard-coded to UTF-8, the call to ZipHelper.RequiresUnicode is worthless and might as well always return true.

>We can take that discussion to a separate issue.

Good idea. I modified the approach of this solution to be a bit more robust and rebased to one commit. The behavior while writing should more closely match desktop now.

ericstj commented:
LGTM

ianhays commented:
Thanks!
